---
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r goi_libs, echo=FALSE, results='hide', include=FALSE}
library(dplyr)
library(knitr)
library(tibble)
library(tidyr)

library(stringr)
library(plotly)
library(DT)
library(corrplot)
library(biomaRt)
library(here)

source(here('munge.R'))
source(here('visualization.R'))

```

## Genes of interest

Batch-corrected transcripts per million (bcTPM) were generated using methods adapted from [Wang et al. 2018](https://www.nature.com/articles/sdata201861) to enable comparison between PACT, [TCGA](https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga), and [GTEX](https://gtexportal.org).

Somatic mutations and copy number were derived from PACT's EPIC pipeline.

[Z-scores](https://www.simplypsychology.org/z-score.html) are the bcTPMs standardized to enable comparisons in expression between genes. If the z-score for a gene is zero, its expression is equal to the mean, while if it is positive or negative, its expression is higher or lower than the mean respectively. A z-score of 1 or -1, for example, means the expression of the gene for that sample is one standard deviation above or below the mean respectively. 

Percentiles are the % ranking relative to other samples from PACT, TCGA, or GTEX and range from 0 to 1.

### Reporting Tables

```{r goi_config, eval=FALSE}
cn_f <- '../test_data/PACT056_T_196454/PACT056_T_196454_segments.txt'
vars_f <- '../test_data/PACT056_T_196454/PACT056_T_196454_baseline_oncotator_results_Annotated.tsv'
```

```{r goi_datain}
dat <- emat[rownames(emat) %in% goi$gene, ]

pact <- dat[, grep('^PACT', names(dat))]
pact <- pact[, names(pact)!=sid]
tcga <- dat[, grep('^TCGA', names(dat))]
gtex <- dat[, grep('^GTEX', names(dat))]

esample=dat[, grep(sid, names(dat))]

```

```{r goi_tab_munge}
rs_dat=c()
zpp=c()
zgp=c()
ztp=c()

for (i in 1:dim(dat)[1]){
    p2=as.numeric(pact[i,])
    t2=as.numeric(tcga[i,])
    g2=as.numeric(gtex[i,])
    s2=esample[i]
    
    # Add sample to end of PACT, TCGA, GTEX values
    p2=c(p2,s2)
    t2=c(t2,s2)
    g2=c(g2,s2)
    zp=zscore(p2)
    zt=zscore(t2)
    zg=zscore(g2)
    
    zps=tail(zp, n=1)
    zts=tail(zt, n=1)
    zgs=tail(zg, n=1)
    if(!is.na(zps)){
       zpp=tail(percentile(zp, zp), n=1)
       ztp=tail(percentile(zt, zt), n=1)
       zgp=tail(percentile(zg, zg), n=1)
    } else {
      zpp=NA
      ztp=NA
      zpg=NA
    }

    ens_gene=row.names(dat)[i]
    gene_info <- goi %>% 
      filter(gene==ens_gene) %>%
      dplyr::select(gene, alias, category)
    
    my_row=c()
    my_tpm=2^s2
   
    rs_row=cbind(gene_info,my_tpm,s2,zps,zts,zgs,zpp,ztp,zgp)
    rs_dat=rbind(rs_dat,rs_row)
}

names(rs_dat)=c("Gene","Alias","Category","bcTPM","log2(bcTPM)","Z-score PACT","Z-score TCGA","Z-score GTEX","Percentile PACT","Percentile TCGA","Percentile GTEX")
```

```{r goi_cnmut}
cn_o <- read.delim(cn_f)
cn <- cn_o

vars_o <- read.delim(vars_f)
vars_cols = c('Hugo_Symbol', 'Protein_Change')
vars <- vars_o %>%
  mutate(Chromosome=paste0('chr',Chromosome)) %>%
  filter(!is.na(Protein_Change)) %>%
  dplyr::select(all_of(vars_cols)) %>%
  distinct()

# Get gene positions for intersect
ens <- useEnsembl(biomart='genes', dataset='hsapiens_gene_ensembl', GRCh=37)#, verbose=TRUE)
ens_canonical <- c(seq(1:22), 'X', 'Y', 'MT')
bm <- getBM(attributes=c('hgnc_symbol', 'chromosome_name', 'start_position', 'end_position'),
            filters='hgnc_symbol', mart=ens, 
            values=genes) %>%
  filter(chromosome_name %in% ens_canonical) %>%
  mutate(chromosome=str_replace(chromosome_name, 'MT', 'M'),
         chromosome = paste0('chr', chromosome)
         ) %>%
  left_join(vars, by=c('hgnc_symbol'='Hugo_Symbol'))

## Hotspots
# Hotspot vars that match protein change
hotspots_pc <- hotspots %>%
  filter(report_any_mutation == 'False') %>%
  dplyr::select(hgnc_symbol, Protein_Change)
bm_hot_pc <- bm %>% inner_join(hotspots_pc)

# Hotspot vars where all mutations are to be reported
hotspots_any <- hotspots %>% 
  filter(report_any_mutation == 'True') %>%
  dplyr::select(hgnc_symbol, Protein_Change)
bm_hot_any <- bm %>%
  filter(hgnc_symbol %in% hotspots_any$hgnc_symbol,
         !is.na(Protein_Change)
  )

bm_hot_all <- rbind(bm_hot_pc, bm_hot_any)
keep <- c(goi_o$gene, bm_hot_all$hgnc_symbol)
bm_final <- bm %>% filter(hgnc_symbol %in% keep)

# Intersect copy number and vars
cnRanges <- GenomicRanges::makeGRangesFromDataFrame(cn, seqnames.field='chromosome', 
        start.field="start.pos", end.field="end.pos", keep.extra.columns=TRUE)
gRanges = GenomicRanges::makeGRangesFromDataFrame(bm_final, seqnames.field='chromosome', 
        start.field="start_position", end.field="end_position", keep.extra.columns=TRUE)

g_m <- findoverlaps_metadata(gRanges, cnRanges) %>%
  dplyr::rename(Gene=hgnc_symbol) %>%
  dplyr::select(Gene, "Mutation (AA)"=Protein_Change, CNt)

```

```{r goi_merge_dat}
cols = c("Gene","Alias","Category", "Mutation (AA)", "CNt", "bcTPM","log2(bcTPM)","Z-score PACT","Z-score TCGA","Z-score GTEX","Percentile PACT","Percentile TCGA","Percentile GTEX")
goi_dat <- g_m %>%
  left_join(rs_dat) %>%
  dplyr::select(all_of(cols))

```

Hotspot somatic mutations are from [Chang et al. 2017](https://cancerdiscovery.aacrjournals.org/content/8/2/174)
```{r goi_hotspot_DT}
hotspot_tab <- bm_hot_all %>%
  inner_join(vars_o, by=c('hgnc_symbol' = 'Hugo_Symbol', 'Protein_Change')) %>%
  dplyr::select(Gene=hgnc_symbol, `Mutation (CDS)`=cDNA_Change, `Mutation (AA)`=Protein_Change,`Class`=Variant_Classification, CNt=CN_total)
hotspot_tab$Gene <- sapply(hotspot_tab[['Gene']], link_gene_string)
  
title = dt_caption("Hotspot somatic mutations and their copy number states (CNt)")
hotspot_tab %>%
  datatable(
    caption=title,
    extensions = dt_ext,
    options = dt_opts,
    rownames=FALSE,
    escape = FALSE
)

```

```{r goi_goi_DT}

round1 <- c("bcTPM","log2(bcTPM)","Z-score PACT","Z-score TCGA","Z-score GTEX")
round2 <- c("Percentile PACT","Percentile TCGA","Percentile GTEX")

goi_tab <- goi_dat %>%
  filter(!is.na(`Z-score PACT`),
         !is.na(`Percentile PACT`)
  ) %>%
  mutate_at(round1, round, 1) %>%
  mutate_at(round2, round, 2)

caption = 'Somatic mutations, tumor copy number (CNt), and batch-corrected (bcTPM) expression of key genes.'
dt_quantile_ramp(goi_tab, c('Percentile PACT', 'Percentile TCGA'), caption)

```

### Boxplots {.tabset}

####  Ag presentation

```{r goi_bar_munge,  echo=FALSE, warning=FALSE, message=FALSE}
dat <- df_goi %>% 
  mutate(Source = if_else(specimen_id == sid, sid, as.character(study))) %>%
  mutate(Source = factor(Source, levels=c(sid, "PACT", "TCGA", "GTEX"))) %>%
  mutate(checkpoint2 = factor(checkpoint2, levels=c('Stimulatory', 'Inhibitory', 'Other')))
         
```

```{r goi_agg_presentation, echo=FALSE, warning=FALSE, message=FALSE}
##get a list
df <- dat %>%
  filter(plot_cat == 'Ag presentation') 

fig <- boxplot_goi(df)
fig
```

####  Receptors

```{r goi_Receptor}
df <- dat %>%
  filter(plot_cat == 'Receptor')

fig <- boxplot_goi(df)
fig
```

####  Co-inhibitor

```{r goi_co-inhibitor}
df <- dat %>%
  filter(plot_cat == 'Coinhibitor')

fig <- boxplot_goi(df)
fig
```

####  Co-stimulator

```{r goi_Co-stimulator}
df <- dat %>%
  filter(plot_cat == 'Costimulator')

fig <- boxplot_goi(df)
fig
```
####  Chemokines

```{r goi_Chemokine}
df <- dat %>%
  filter(plot_cat == 'Chemokine')

fig <- boxplot_goi(df)
fig
```

####  Chemokine receptors

```{r goi_Chemokine_receptors}
df <- dat %>%
  filter(plot_cat == 'ChemokineR')

fig <- boxplot_goi(df)
fig
```


#### IFNG

```{r goi_ifng}
df <- dat %>%
  filter(plot_cat == 'IFNG')

fig <- boxplot_goi(df)
fig
```

#### TGFB

```{r goi_tgfb}
df <- dat %>%
  filter(plot_cat == 'TGFB')

fig <- boxplot_goi(df)
fig
```

####  Other

```{r goi_other}
df <- dat %>%
  filter(plot_cat == 'Other')

fig <- boxplot_goi(df)
fig
```

### {-}

### Correlation Matrix {.tabset .tabset-fade}

Correlations among key genes for `r pt_dat[['TCGA Study Code']]` PACT samples (n=`r sum(grepl('^PACT', colnames(emat)))`).

#### GOI

```{r goi_corrplot_goi}
shortlist <- goi %>%
  filter(!is.na(priority),
         !grepl('Chemokine', category)) %>%
  pull(gene) 

dat <- emat[rownames(emat) %in% shortlist, grepl('^PACT', colnames(emat))]
novar <- apply(dat, 1, var) == 0
dat <- dat[!novar, ]
xcor <- cor(t(dat))
corrplot(xcor, type = 'upper', method = 'square', order='hclust', hclust.method='ward.D2', tl.col='black', tl.cex=0.75)
```

#### Chemokines
```{r goi_corrplot_chemokine}
shortlist <- goi %>%
  filter(
         grepl('Chemokine', category)) %>%
  pull(gene) 

dat <- emat[rownames(emat) %in% shortlist, grepl('^PACT', colnames(emat))]
novar <- apply(dat, 1, var) == 0
dat <- dat[!novar, ]
xcor <- cor(t(dat))
corrplot(xcor, type = 'upper', method = 'square', order='hclust', hclust.method='ward.D2', tl.col='black', tl.cex=0.60)
```

## {-}
