

<!DOCTYPE html>
<html class="writer-html4" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Data processing using HISAT2 and HISAT-genotype &mdash; hlahat alpha documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Resources" href="resources.html" />
    <link rel="prev" title="Usage" href="usage.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> hlahat
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Data processing using HISAT2 and HISAT-genotype</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#hla-read-extraction">HLA read extraction</a></li>
<li class="toctree-l2"><a class="reference internal" href="#hla-typing">HLA typing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#constructing-a-custom-hla-reference">Constructing a custom HLA reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#variant-detection">Variant detection</a></li>
<li class="toctree-l2"><a class="reference internal" href="#quantifying-allelic-imbalance">Quantifying allelic imbalance</a></li>
<li class="toctree-l2"><a class="reference internal" href="#paired-tumor-normal-data">Paired tumor-normal data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#tumor-only-datasets">Tumor-only datasets</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="resources.html">Resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">hlahat</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Data processing using HISAT2 and HISAT-genotype</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/analysis.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="data-processing-using-hisat2-and-hisat-genotype">
<h1>Data processing using HISAT2 and HISAT-genotype<a class="headerlink" href="#data-processing-using-hisat2-and-hisat-genotype" title="Permalink to this headline">¶</a></h1>
<div class="section" id="hla-read-extraction">
<h2>HLA read extraction<a class="headerlink" href="#hla-read-extraction" title="Permalink to this headline">¶</a></h2>
<p>Extract HLA-associated reads:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>export PATH=/opt/hisat2/hisat2-hisat2_v2.2.0_beta:/opt/hisat2/hisat2-hisat2_v2.2.0_beta/hisatgenotype_scripts:$PATH
export PYTHONPATH=/opt/hisat2/hisat2-hisat2_v2.2.0_beta/hisatgenotype_modules:$PYTHONPATH
export PATH=$PATH:/opt/samtools/bin
/opt/hisat2/hisat2-hisat2_v2.2.0_beta/hisatgenotype_extract_reads_v_KC.py --base ${hisat_prefix} \
        -1 ${fq1} -2 ${fq2} ${true=&#39;--is-rna&#39; false=&#39;&#39; is_rna}\
        --database-list hla
</pre></div>
</div>
</div>
<div class="section" id="hla-typing">
<h2>HLA typing<a class="headerlink" href="#hla-typing" title="Permalink to this headline">¶</a></h2>
<p>Docker Commands (kcampbel/rnaseq_methods:v3):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>export PATH=/opt/hisat2/hisat2-hisat2_v2.2.0_beta:/opt/hisat2/hisat2-hisat2_v2.2.0_beta/hisatgenotype_scripts:$PATH
export PYTHONPATH=/opt/hisat2/hisat2-hisat2_v2.2.0_beta/hisatgenotype_modules:$PYTHONPATH
export PATH=$PATH:/opt/samtools/bin
/opt/hisat2/hisat2-hisat2_v2.2.0_beta/hisatgenotype_locus_v_KC.py --genotype-genome ${hisat_prefix} \
        -U ${fq1} ${true=&#39;--is-rna&#39; false=&#39;&#39; is_rna}\
        --base hla --locus-list ${locus} --output-base ${name}.${locus} --keep-low-abundance-alleles
</pre></div>
</div>
<p>Documentation for HISAT-genotype suggests using the alleles ranked 1 or 2, from hisatgenotype_locus.py for each gene, and the alleles may be reported up to the 4th field of resolution, which describes genomic differences in alleles outside of the coding regions. However, WES may not have sufficient sequencing coverage and RNAseq data would not be appropriate for detecting this level of information. HLA-HAT outputs the ${id}.all_types.tsv file, indicating the ranked alleles, by abundance, to include the most comprehensive output from HISAT-genotype.
${id}.all_types.tsv file is a tab-delimited file derived from the report outputted by hisatgenotype_locus.py:</p>
<table border="1" class="colwidths-auto docutils align-center" id="id1">
<caption><span class="caption-text">all_types</span><a class="headerlink" href="#id1" title="Permalink to this table">¶</a></caption>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Field</th>
<th class="head">Type</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>ranks</td>
<td>Integer</td>
<td>Gene rank of allele, based upon percent abundance of reads assigned to corresponding HLA type</td>
</tr>
<tr class="row-odd"><td>alleles</td>
<td>String</td>
<td>Full resolution of ranked allele identified by HISAT-genotype</td>
</tr>
<tr class="row-even"><td>gene</td>
<td>String</td>
<td>HLA gene</td>
</tr>
<tr class="row-odd"><td>perc_abundance</td>
<td>Float</td>
<td>Relative abundance of reads corresponding to allele</td>
</tr>
</tbody>
</table>
<p>By default, all alleles are reduced to their fullest resolution or up the third field of resolution (e.g. A*02:89 would remain A*02:89, while A*03:01:01:01 is reduced to A*03:01:01). Then, alleles up to the third field of resolution are summarized by the maximum percent abundance across those that are shared. Any alleles with less than 5% abundance are removed, and then the remaining one or top two alleles (at the third field of resolution) are chosen as the HLA types.</p>
<p><em>Optional</em>: To summarize at the second field of resolution, the flag –field_of_resolution 2 can be used. If this parameter is set, the top two alleles at the second field of resolution are chosen.</p>
<p><em>Example</em>: If the following Class I alleles are ranked in the report from HISAT-genotype:
.. table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">:</span><span class="n">widths</span><span class="p">:</span> <span class="n">auto</span>
      <span class="p">:</span><span class="n">align</span><span class="p">:</span> <span class="n">center</span>
      <span class="p">:</span><span class="n">header</span><span class="o">-</span><span class="n">rows</span><span class="p">:</span> <span class="mi">1</span>

             <span class="o">=====</span>  <span class="o">===============</span>  <span class="o">=====</span>  <span class="o">===============</span>
             <span class="n">ranks</span>  <span class="n">alleles</span>          <span class="n">gene</span>   <span class="n">perc_abundance</span>
             <span class="o">=====</span>  <span class="o">===============</span>  <span class="o">=====</span>  <span class="o">===============</span>
             <span class="mi">1</span>      <span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>    <span class="n">A</span>      <span class="mf">40.85</span>
             <span class="mi">2</span>      <span class="n">A</span><span class="o">*</span><span class="mi">33</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>       <span class="n">A</span>      <span class="mf">31.63</span>
             <span class="mi">3</span>      <span class="n">A</span><span class="o">*</span><span class="mi">33</span><span class="p">:</span><span class="mi">03</span><span class="p">:</span><span class="mi">23</span>       <span class="n">A</span>      <span class="mf">13.97</span>
             <span class="mi">4</span>      <span class="n">A</span><span class="o">*</span><span class="mi">34</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>       <span class="n">A</span>      <span class="mf">4.52</span>
             <span class="mi">5</span>      <span class="n">A</span><span class="o">*</span><span class="mi">34</span><span class="p">:</span><span class="mi">05</span>          <span class="n">A</span>      <span class="mf">4.52</span>
             <span class="mi">6</span>      <span class="n">A</span><span class="o">*</span><span class="mi">34</span><span class="p">:</span><span class="mi">14</span>          <span class="n">A</span>      <span class="mf">4.52</span>
             <span class="mi">1</span>      <span class="n">B</span><span class="o">*</span><span class="mi">14</span><span class="p">:</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>    <span class="n">B</span>      <span class="mf">50.79</span>
             <span class="mi">2</span>      <span class="n">B</span><span class="o">*</span><span class="mi">15</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>    <span class="n">B</span>      <span class="mf">37.33</span>
             <span class="mi">3</span>      <span class="n">B</span><span class="o">*</span><span class="mi">15</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">03</span>    <span class="n">B</span>      <span class="mf">11.87</span>
             <span class="mi">1</span>      <span class="n">C</span><span class="o">*</span><span class="mi">08</span><span class="p">:</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>    <span class="n">C</span>      <span class="mf">51.18</span>
             <span class="mi">2</span>      <span class="n">C</span><span class="o">*</span><span class="mi">03</span><span class="p">:</span><span class="mi">03</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>    <span class="n">C</span>      <span class="mf">48.82</span>
             <span class="o">=====</span> <span class="o">===============</span>   <span class="o">=====</span>  <span class="o">===============</span>
</pre></div>
</div>
<p>First, alleles are summarized to the third field of resolution:
..  table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">:</span><span class="n">align</span><span class="p">:</span> <span class="n">center</span>
<span class="p">:</span><span class="n">header</span><span class="o">-</span><span class="n">rows</span><span class="p">:</span> <span class="mi">1</span>
<span class="p">:</span><span class="n">widths</span><span class="p">:</span> <span class="n">auto</span>

<span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span>
<span class="n">ranks</span> <span class="n">alleles</span> <span class="n">gene</span>  <span class="n">perc_abundance</span>
<span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span>
<span class="mi">1</span> <span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span> <span class="n">A</span> <span class="mf">40.85</span>
<span class="mi">2</span> <span class="n">A</span><span class="o">*</span><span class="mi">33</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>  <span class="n">A</span> <span class="mf">31.63</span>
<span class="mi">3</span> <span class="n">A</span><span class="o">*</span><span class="mi">33</span><span class="p">:</span><span class="mi">03</span><span class="p">:</span><span class="mi">23</span>  <span class="n">A</span> <span class="mf">13.97</span>
<span class="mi">4</span> <span class="n">A</span><span class="o">*</span><span class="mi">34</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>  <span class="n">A</span> <span class="mf">4.52</span>
<span class="mi">5</span> <span class="n">A</span><span class="o">*</span><span class="mi">34</span><span class="p">:</span><span class="mi">05</span> <span class="n">A</span> <span class="mf">4.52</span>
<span class="mi">6</span> <span class="n">A</span><span class="o">*</span><span class="mi">34</span><span class="p">:</span><span class="mi">14</span> <span class="n">A</span> <span class="mf">4.52</span>
<span class="mi">1</span> <span class="n">B</span><span class="o">*</span><span class="mi">14</span><span class="p">:</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span> <span class="n">B</span> <span class="mf">50.79</span>
<span class="mi">2</span> <span class="n">B</span><span class="o">*</span><span class="mi">15</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span> <span class="n">B</span> <span class="mf">37.33</span>
<span class="mi">3</span> <span class="n">B</span><span class="o">*</span><span class="mi">15</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span> <span class="n">B</span> <span class="mf">11.87</span>
<span class="mi">1</span> <span class="n">C</span><span class="o">*</span><span class="mi">08</span><span class="p">:</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span> <span class="n">C</span> <span class="mf">51.18</span>
<span class="mi">2</span> <span class="n">C</span><span class="o">*</span><span class="mi">03</span><span class="p">:</span><span class="mi">03</span><span class="p">:</span><span class="mi">01</span> <span class="n">C</span> <span class="mf">48.82</span>
<span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span>
</pre></div>
</div>
<p>Alleles are summarized by the maximum percent abundance corresponding to each unique allele at the third field of resolution:
..  table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">:</span><span class="n">align</span><span class="p">:</span> <span class="n">center</span>
<span class="p">:</span><span class="n">header</span><span class="o">-</span><span class="n">rows</span><span class="p">:</span> <span class="mi">1</span>
<span class="p">:</span><span class="n">widths</span><span class="p">:</span> <span class="n">auto</span>

<span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span>
<span class="n">ranks</span> <span class="n">alleles</span> <span class="n">gene</span>  <span class="n">perc_abundance</span>
<span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span>
<span class="mi">1</span> <span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span> <span class="n">A</span> <span class="mf">40.85</span>
<span class="mi">2</span> <span class="n">A</span><span class="o">*</span><span class="mi">33</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>  <span class="n">A</span> <span class="mf">31.63</span>
<span class="mi">3</span> <span class="n">A</span><span class="o">*</span><span class="mi">33</span><span class="p">:</span><span class="mi">03</span><span class="p">:</span><span class="mi">23</span>  <span class="n">A</span> <span class="mf">13.97</span>
<span class="mi">4</span> <span class="n">A</span><span class="o">*</span><span class="mi">34</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>  <span class="n">A</span> <span class="mf">4.52</span>
<span class="mi">5</span> <span class="n">A</span><span class="o">*</span><span class="mi">34</span><span class="p">:</span><span class="mi">05</span> <span class="n">A</span> <span class="mf">4.52</span>
<span class="mi">6</span> <span class="n">A</span><span class="o">*</span><span class="mi">34</span><span class="p">:</span><span class="mi">14</span> <span class="n">A</span> <span class="mf">4.52</span>
<span class="mi">1</span> <span class="n">B</span><span class="o">*</span><span class="mi">14</span><span class="p">:</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span> <span class="n">B</span> <span class="mf">50.79</span>
<span class="mi">2</span> <span class="n">B</span><span class="o">*</span><span class="mi">15</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span> <span class="n">B</span> <span class="mf">37.33</span>
<span class="mi">1</span> <span class="n">C</span><span class="o">*</span><span class="mi">08</span><span class="p">:</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span> <span class="n">C</span> <span class="mf">51.18</span>
<span class="mi">2</span> <span class="n">C</span><span class="o">*</span><span class="mi">03</span><span class="p">:</span><span class="mi">03</span><span class="p">:</span><span class="mi">01</span> <span class="n">C</span> <span class="mf">48.82</span>
<span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span>
</pre></div>
</div>
<p>Alleles with less than 5% abundance are removed:
..  table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">:</span><span class="n">align</span><span class="p">:</span> <span class="n">center</span>
<span class="p">:</span><span class="n">header</span><span class="o">-</span><span class="n">rows</span><span class="p">:</span> <span class="mi">1</span>
<span class="p">:</span><span class="n">widths</span><span class="p">:</span> <span class="n">auto</span>

<span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span>
<span class="n">ranks</span> <span class="n">alleles</span> <span class="n">gene</span>  <span class="n">perc_abundance</span>
<span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span>
<span class="mi">1</span> <span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span> <span class="n">A</span> <span class="mf">40.85</span>
<span class="mi">2</span> <span class="n">A</span><span class="o">*</span><span class="mi">33</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>  <span class="n">A</span> <span class="mf">31.63</span>
<span class="mi">3</span> <span class="n">A</span><span class="o">*</span><span class="mi">33</span><span class="p">:</span><span class="mi">03</span><span class="p">:</span><span class="mi">23</span>  <span class="n">A</span> <span class="mf">13.97</span>
<span class="mi">1</span> <span class="n">B</span><span class="o">*</span><span class="mi">14</span><span class="p">:</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span> <span class="n">B</span> <span class="mf">50.79</span>
<span class="mi">2</span> <span class="n">B</span><span class="o">*</span><span class="mi">15</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span> <span class="n">B</span> <span class="mf">37.33</span>
<span class="mi">1</span> <span class="n">C</span><span class="o">*</span><span class="mi">08</span><span class="p">:</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span> <span class="n">C</span> <span class="mf">51.18</span>
<span class="mi">2</span> <span class="n">C</span><span class="o">*</span><span class="mi">03</span><span class="p">:</span><span class="mi">03</span><span class="p">:</span><span class="mi">01</span> <span class="n">C</span> <span class="mf">48.82</span>
<span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span>
</pre></div>
</div>
<p>Finally, the top 1-2 ranked alleles are identified as the patient HLA type:
..  table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">:</span><span class="n">align</span><span class="p">:</span> <span class="n">center</span>
<span class="p">:</span><span class="n">header</span><span class="o">-</span><span class="n">rows</span><span class="p">:</span> <span class="mi">1</span>
<span class="p">:</span><span class="n">widths</span><span class="p">:</span> <span class="n">auto</span>

<span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span>
<span class="n">ranks</span> <span class="n">alleles</span> <span class="n">gene</span>  <span class="n">perc_abundance</span>
<span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span>
<span class="mi">1</span> <span class="n">A</span><span class="o">*</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span> <span class="n">A</span> <span class="mf">40.85</span>
<span class="mi">2</span> <span class="n">A</span><span class="o">*</span><span class="mi">33</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span>  <span class="n">A</span> <span class="mf">31.63</span>
<span class="mi">1</span> <span class="n">B</span><span class="o">*</span><span class="mi">14</span><span class="p">:</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span> <span class="n">B</span> <span class="mf">50.79</span>
<span class="mi">2</span> <span class="n">B</span><span class="o">*</span><span class="mi">15</span><span class="p">:</span><span class="mi">01</span><span class="p">:</span><span class="mi">01</span> <span class="n">B</span> <span class="mf">37.33</span>
<span class="mi">1</span> <span class="n">C</span><span class="o">*</span><span class="mi">08</span><span class="p">:</span><span class="mi">02</span><span class="p">:</span><span class="mi">01</span> <span class="n">C</span> <span class="mf">51.18</span>
<span class="mi">2</span> <span class="n">C</span><span class="o">*</span><span class="mi">03</span><span class="p">:</span><span class="mi">03</span><span class="p">:</span><span class="mi">01</span> <span class="n">C</span> <span class="mf">48.82</span>
<span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span> <span class="o">=====</span>
</pre></div>
</div>
<p>The final list of HLA types is summarized by ${id}.top_hlatypes.tsv, a tab-delimited file containing the filtered allele calls:
.. rubric::top_hlatypes File Format
.. list-table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">:</span><span class="n">align</span><span class="p">:</span> <span class="n">center</span>
<span class="p">:</span><span class="n">header</span><span class="o">-</span><span class="n">rows</span><span class="p">:</span> <span class="mi">1</span>

<span class="o">*</span> <span class="o">-</span> <span class="n">Field</span>
        <span class="o">-</span> <span class="n">Type</span>
        <span class="o">-</span> <span class="n">Description</span>
<span class="o">*</span> <span class="o">-</span> <span class="n">gene</span>
        <span class="o">-</span> <span class="n">String</span>
        <span class="o">-</span> <span class="n">HLA</span> <span class="n">gene</span>
<span class="o">*</span> <span class="o">-</span> <span class="n">allele</span>
        <span class="o">-</span> <span class="n">String</span>
        <span class="o">-</span> <span class="n">Filtered</span> <span class="n">allele</span> <span class="n">call</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="constructing-a-custom-hla-reference">
<h1>Constructing a custom HLA reference<a class="headerlink" href="#constructing-a-custom-hla-reference" title="Permalink to this headline">¶</a></h1>
<p>docker: kcampbel/hlahat_r:v1</p>
<blockquote>
<div>grep “ranked” ${sep=” ” hla_report_files} &gt; ${name}.hla_types.txt
Rscript /code/generate_reference_files.R ${name} ${hlatypes} ${sep=”,” gen_msf_list} ${sep=”,” nuc_msf_list}</div></blockquote>
<div class="section" id="variant-detection">
<h2>Variant detection<a class="headerlink" href="#variant-detection" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="quantifying-allelic-imbalance">
<h2>Quantifying allelic imbalance<a class="headerlink" href="#quantifying-allelic-imbalance" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="paired-tumor-normal-data">
<h2>Paired tumor-normal data<a class="headerlink" href="#paired-tumor-normal-data" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="tumor-only-datasets">
<h2>Tumor-only datasets<a class="headerlink" href="#tumor-only-datasets" title="Permalink to this headline">¶</a></h2>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="resources.html" class="btn btn-neutral float-right" title="Resources" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="usage.html" class="btn btn-neutral float-left" title="Usage" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Katie Campbell.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>