

<!DOCTYPE html>
<html class="writer-html4" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Analysis &mdash; hlahat alpha documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Resources" href="resources.html" />
    <link rel="prev" title="Usage" href="usage.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> hlahat
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Analysis</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-1-preparation-and-assignment">Module 1. Preparation and assignment</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#hla-read-extraction">HLA read extraction</a></li>
<li class="toctree-l3"><a class="reference internal" href="#hla-typing">HLA typing</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-2-reference-generation">Module 2. Reference Generation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#hla-haplotype-reporting">HLA haplotype reporting</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-3-downstream-analysis">Module 3. Downstream Analysis</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#paired-tumor-normal-data">Paired tumor-normal data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#tumor-only-datasets">Tumor-only datasets</a></li>
<li class="toctree-l3"><a class="reference internal" href="#quantifying-allelic-imbalance">Quantifying allelic imbalance</a></li>
<li class="toctree-l3"><a class="reference internal" href="#variant-detection">Variant detection</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="resources.html">Resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">hlahat</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Analysis</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/analysis.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="analysis">
<h1>Analysis<a class="headerlink" href="#analysis" title="Permalink to this headline">¶</a></h1>
<a class="reference internal image-reference" href="_images/workflow.jpg"><img alt="_images/workflow.jpg" class="align-center" src="_images/workflow.jpg" style="width: 800px;" /></a>
<div class="section" id="module-1-preparation-and-assignment">
<h2>Module 1. Preparation and assignment<a class="headerlink" href="#module-1-preparation-and-assignment" title="Permalink to this headline">¶</a></h2>
<div class="section" id="hla-read-extraction">
<h3>HLA read extraction<a class="headerlink" href="#hla-read-extraction" title="Permalink to this headline">¶</a></h3>
<p>Reads associated with the HLA gene loci are extracted using HISAT-genotype. The <code class="docutils literal notranslate"><span class="pre">hisatgenotype_extract_reads</span></code> file has been modified to work with either DNA or RNA sequencing reads (enabling split-read sequence alignment through HISAT2).</p>
<p>Extract HLA-associated reads from DNA sequencing data:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>export PATH=/opt/hisat2/hisat2-hisat2_v2.2.0_beta:/opt/hisat2/hisat2-hisat2_v2.2.0_beta/hisatgenotype_scripts:$PATH
export PYTHONPATH=/opt/hisat2/hisat2-hisat2_v2.2.0_beta/hisatgenotype_modules:$PYTHONPATH
export PATH=$PATH:/opt/samtools/bin
/opt/hisat2/hisat2-hisat2_v2.2.0_beta/hisatgenotype_extract_reads_v_KC.py --base ${hisat_prefix} \
-1 ${fq1} -2 ${fq2} --database-list hla
</pre></div>
</div>
<p>To extract HLA-associated reads from RNA sequencing data, just add the <code class="docutils literal notranslate"><span class="pre">--is-rna</span></code> flag to the command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>export PATH=/opt/hisat2/hisat2-hisat2_v2.2.0_beta:/opt/hisat2/hisat2-hisat2_v2.2.0_beta/hisatgenotype_scripts:$PATH
export PYTHONPATH=/opt/hisat2/hisat2-hisat2_v2.2.0_beta/hisatgenotype_modules:$PYTHONPATH
export PATH=$PATH:/opt/samtools/bin
/opt/hisat2/hisat2-hisat2_v2.2.0_beta/hisatgenotype_extract_reads_v_KC.py --base ${hisat_prefix} \
-1 ${fq1} -2 ${fq2} --is-rna --database-list hla
</pre></div>
</div>
<p>Reads may also be extracted from single reads, using the <code class="docutils literal notranslate"><span class="pre">-U</span></code> option instead of paired <code class="docutils literal notranslate"><span class="pre">-1</span></code> and <code class="docutils literal notranslate"><span class="pre">-2</span></code> reads:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>export PATH=/opt/hisat2/hisat2-hisat2_v2.2.0_beta:/opt/hisat2/hisat2-hisat2_v2.2.0_beta/hisatgenotype_scripts:$PATH
export PYTHONPATH=/opt/hisat2/hisat2-hisat2_v2.2.0_beta/hisatgenotype_modules:$PYTHONPATH
export PATH=$PATH:/opt/samtools/bin
/opt/hisat2/hisat2-hisat2_v2.2.0_beta/hisatgenotype_extract_reads_v_KC.py --base ${hisat_prefix} \
-U ${fq1} --is-rna --database-list hla
</pre></div>
</div>
</div>
<div class="section" id="hla-typing">
<h3>HLA typing<a class="headerlink" href="#hla-typing" title="Permalink to this headline">¶</a></h3>
<p>HLA typing can be applied to 26 genes, spanning 7Mb in chromosome 6p21.3, including classical and non-classical Class I/II HLA genes, non-expressed Class I HLA pseudogenes, ATP binding cassette transporter genes, Class I chain-related and Class I-like genes: <em>HLA-A, HLA-B, HLA-C, HLA-DMA, HLA-DMB, HLA-DOA, HLA-DOB, HLA-DPA1, HLA-DPB1, HLA-DPB2, HLA-DQA1, HLA-DQB1, HLA-DRA, HLA-DRB1, HLA-E, HLA-F, HLA-G, HLA-H, HFE, HLA-K, HLA-L, MICA, MICB, TAP1, TAP2,</em> and <em>HLA-V</em>. These are specified as a comma-delimited list in the following HLA typing command by the <cite>–locus-list</cite> option.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Do not include “HLA-” when specifying HLA genes (e.g. use <code class="docutils literal notranslate"><span class="pre">--locus-list</span> <span class="pre">A</span></code> instead of <code class="docutils literal notranslate"><span class="pre">--locus-list</span> <span class="pre">HLA-A</span></code>). Typing generally takes the longest on the HLA-A locus, and this command can be scattered across each loci individually, followed by report aggregation and summary (See suggested workflow).</p>
</div>
<p>Docker Commands (kcampbel/rnaseq_methods:v3):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>export PATH=/opt/hisat2/hisat2-hisat2_v2.2.0_beta:/opt/hisat2/hisat2-hisat2_v2.2.0_beta/hisatgenotype_scripts:$PATH
export PYTHONPATH=/opt/hisat2/hisat2-hisat2_v2.2.0_beta/hisatgenotype_modules:$PYTHONPATH
export PATH=$PATH:/opt/samtools/bin
/opt/hisat2/hisat2-hisat2_v2.2.0_beta/hisatgenotype_locus_v_KC.py --genotype-genome ${hisat_prefix} \
-1 ${fq1} -- ${fq2} --base hla --locus-list ${locus} --output-base ${name}.${locus} --keep-low-abundance-alleles
</pre></div>
</div>
<p>Options for single-read sequencing (<code class="docutils literal notranslate"><span class="pre">U</span></code>) and RNA sequencing (<code class="docutils literal notranslate"><span class="pre">--is-rna</span></code>) can also be used for HLA typing.</p>
</div>
</div>
<div class="section" id="module-2-reference-generation">
<h2>Module 2. Reference Generation<a class="headerlink" href="#module-2-reference-generation" title="Permalink to this headline">¶</a></h2>
<p>HISAT-genotype will output a ranked list of HLA types for each locus, based upon those that meet the expectation maximization algorithm. Documentation for HISAT-genotype suggests using the alleles ranked 1 or 2, from hisatgenotype_locus.py for each gene, and the alleles may be reported up to the 4th field of resolution, which describes genomic differences in alleles outside of the coding regions. However, WES may not have sufficient sequencing coverage and RNAseq data would not be appropriate for detecting this level of information. HLA-HAT provides an R script to summarize the HLA typing reports from HISAT-genotype.</p>
<div class="section" id="hla-haplotype-reporting">
<h3>HLA haplotype reporting<a class="headerlink" href="#hla-haplotype-reporting" title="Permalink to this headline">¶</a></h3>
<p>Aggregate all report files using the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>grep &quot;ranked&quot; *report &gt; ${name}.hla_types.txt
</pre></div>
</div>
<p>The generate_reference_files.R script is used to generate the reference fasta, docker: kcampbel/hlahat_r:v3:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Rscript /code/generate_reference_files.R ${name} ${hlatypes} ${n_fields} ${sep=&quot;,&quot; gen_msf_list} ${sep=&quot;,&quot; nuc_msf_list}
</pre></div>
</div>
<p>The R script <code class="docutils literal notranslate"><span class="pre">generate_reference_files.R</span></code> is provided for summarizing the HLA typing performed by HISAT-genotype, and takes the following arguments:</p>
<p>HLA-HAT outputs the <em>${id}.all_types.tsv</em> file, indicating the ranked alleles, by abundance, to include the most comprehensive output from HISAT-genotype.</p>
<p><em>${id}.all_types.tsv</em> file is a tab-delimited file derived from the report outputted by hisatgenotype_locus.py:</p>
<table border="1" class="colwidths-auto docutils align-center">
<thead valign="bottom">
<tr class="row-odd"><th class="head">Column Name</th>
<th class="head">Type</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>ranks</td>
<td>Integer</td>
<td>Gene rank of allele, based upon percent abundance of reads assigned to corresponding HLA type</td>
</tr>
<tr class="row-odd"><td>gene</td>
<td>String</td>
<td>HLA gene</td>
</tr>
<tr class="row-even"><td>perc_abundance</td>
<td>Float</td>
<td>Relative abundance of reads corresponding to allele</td>
</tr>
</tbody>
</table>
<p>By default, all alleles are reduced to their fullest resolution or up the third field of resolution (e.g. A*02:89 would remain A*02:89, while A*03:01:01:01 is reduced to A*03:01:01). Then, alleles up to the third field of resolution are summarized by the maximum percent abundance across those that are shared. Any alleles with less than 5% abundance are removed, and then the remaining one or top two alleles (at the third field of resolution) are chosen as the HLA types.</p>
<p><em>Optional</em>: To summarize at the second field of resolution, the flag <code class="docutils literal notranslate"><span class="pre">--field_of_resolution</span> <span class="pre">2</span></code> can be used. If this parameter is set, the top two alleles at the second field of resolution are chosen.</p>
<p><em>Example</em>: If the following Class I alleles are ranked in the report from HISAT-genotype:
.. csv-table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span> <span class="p">:</span><span class="n">widths</span><span class="p">:</span> <span class="n">auto</span>
 <span class="p">:</span><span class="n">align</span><span class="p">:</span> <span class="n">center</span>
 <span class="p">:</span><span class="n">header</span><span class="p">:</span> <span class="s2">&quot;ranks&quot;</span><span class="p">,</span> <span class="s2">&quot;alleles&quot;</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;perc_abundance&quot;</span>

<span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;A*02:01:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;40.85&quot;</span>
<span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;A*33:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;31.63&quot;</span>
<span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="s2">&quot;A*33:03:23&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;13.97&quot;</span>
<span class="s2">&quot;4&quot;</span><span class="p">,</span> <span class="s2">&quot;A*34:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;4.52&quot;</span>
<span class="s2">&quot;5&quot;</span><span class="p">,</span> <span class="s2">&quot;A*34:05&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;4.52&quot;</span>
<span class="s2">&quot;6&quot;</span><span class="p">,</span> <span class="s2">&quot;A*34:14&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;4.52&quot;</span>
<span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;B*14:02:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;50.79&quot;</span>
<span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;B*15:01:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;37.33&quot;</span>
<span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="s2">&quot;B*15:01:01:03&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;11.87&quot;</span>
<span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;C*08:02:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;51.18&quot;</span>
<span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;C*03:03:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;48.82&quot;</span>
</pre></div>
</div>
<p>First, alleles are summarized to the third field of resolution:
.. csv-table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">:</span><span class="n">widths</span><span class="p">:</span> <span class="n">auto</span>
<span class="p">:</span><span class="n">align</span><span class="p">:</span> <span class="n">center</span>
<span class="p">:</span><span class="n">header</span><span class="p">:</span> <span class="s2">&quot;ranks&quot;</span><span class="p">,</span> <span class="s2">&quot;alleles&quot;</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;perc_abundance&quot;</span>

<span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;A*02:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;40.85&quot;</span>
<span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;A*33:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;31.63&quot;</span>
<span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="s2">&quot;A*33:03:23&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;13.97&quot;</span>
<span class="s2">&quot;4&quot;</span><span class="p">,</span> <span class="s2">&quot;A*34:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;4.52&quot;</span>
<span class="s2">&quot;5&quot;</span><span class="p">,</span> <span class="s2">&quot;A*34:05&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;4.52&quot;</span>
<span class="s2">&quot;6&quot;</span><span class="p">,</span> <span class="s2">&quot;A*34:14&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;4.52&quot;</span>
<span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;B*14:02:01&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;50.79&quot;</span>
<span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;B*15:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;37.33&quot;</span>
<span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="s2">&quot;B*15:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;11.87&quot;</span>
<span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;C*08:02:01&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;51.18&quot;</span>
<span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;C*03:03:01&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;48.82&quot;</span>
</pre></div>
</div>
<p>Alleles are summarized by the maximum percent abundance corresponding to each unique allele at the third field of resolution:
.. csv-table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">:</span><span class="n">widths</span><span class="p">:</span> <span class="n">auto</span>
<span class="p">:</span><span class="n">align</span><span class="p">:</span> <span class="n">center</span>
<span class="p">:</span><span class="n">header</span><span class="p">:</span> <span class="s2">&quot;ranks&quot;</span><span class="p">,</span> <span class="s2">&quot;alleles&quot;</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;perc_abundance&quot;</span>

<span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;A*02:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;40.85&quot;</span>
<span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;A*33:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;31.63&quot;</span>
<span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="s2">&quot;A*33:03:23&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;13.97&quot;</span>
<span class="s2">&quot;4&quot;</span><span class="p">,</span> <span class="s2">&quot;A*34:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;4.52&quot;</span>
<span class="s2">&quot;5&quot;</span><span class="p">,</span> <span class="s2">&quot;A*34:05&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;4.52&quot;</span>
<span class="s2">&quot;6&quot;</span><span class="p">,</span> <span class="s2">&quot;A*34:14&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;4.52&quot;</span>
<span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;B*14:02:01&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;50.79&quot;</span>
<span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;B*15:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;37.33&quot;</span>
<span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;C*08:02:01&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;51.18&quot;</span>
<span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;C*03:03:01&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;48.82&quot;</span>
</pre></div>
</div>
<p>Alleles with less than 5% abundance are removed:
.. csv-table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">:</span><span class="n">widths</span><span class="p">:</span> <span class="n">auto</span>
<span class="p">:</span><span class="n">align</span><span class="p">:</span> <span class="n">center</span>
<span class="p">:</span><span class="n">header</span><span class="p">:</span> <span class="s2">&quot;ranks&quot;</span><span class="p">,</span> <span class="s2">&quot;alleles&quot;</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;perc_abundance&quot;</span>

<span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;A*02:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;40.85&quot;</span>
<span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;A*33:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;31.63&quot;</span>
<span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="s2">&quot;A*33:03:23&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;13.97&quot;</span>
<span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;B*14:02:01&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;50.79&quot;</span>
<span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;B*15:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;37.33&quot;</span>
<span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;C*08:02:01&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;51.18&quot;</span>
<span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;C*03:03:01&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;48.82&quot;</span>
</pre></div>
</div>
<p>Finally, the top 1-2 ranked alleles are identified as the patient HLA type:
.. csv-table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">:</span><span class="n">widths</span><span class="p">:</span> <span class="n">auto</span>
<span class="p">:</span><span class="n">align</span><span class="p">:</span> <span class="n">center</span>
<span class="p">:</span><span class="n">header</span><span class="p">:</span> <span class="s2">&quot;ranks&quot;</span><span class="p">,</span> <span class="s2">&quot;alleles&quot;</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;perc_abundance&quot;</span>

<span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;A*02:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;40.85&quot;</span>
<span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;A*33:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;31.63&quot;</span>
<span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;B*14:02:01&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;50.79&quot;</span>
<span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;B*15:01:01&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;37.33&quot;</span>
<span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;C*08:02:01&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;51.18&quot;</span>
<span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;C*03:03:01&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;48.82&quot;</span>
</pre></div>
</div>
<p>The final list of HLA types is summarized by <em>${id}.top_hlatypes.tsv</em>, a tab-delimited file containing the filtered allele calls:
.. csv-table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">:</span><span class="n">widths</span><span class="p">:</span> <span class="n">auto</span>
<span class="p">:</span><span class="n">align</span><span class="p">:</span> <span class="n">center</span>
<span class="p">:</span><span class="n">header</span><span class="p">:</span> <span class="s2">&quot;Field&quot;</span><span class="p">,</span> <span class="s2">&quot;Type&quot;</span><span class="p">,</span> <span class="s2">&quot;Description&quot;</span>

<span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="s2">&quot;HLA gene&quot;</span>
<span class="s2">&quot;allele&quot;</span><span class="p">,</span> <span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="s2">&quot;Filtered allele call&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="module-3-downstream-analysis">
<h2>Module 3. Downstream Analysis<a class="headerlink" href="#module-3-downstream-analysis" title="Permalink to this headline">¶</a></h2>
<div class="section" id="paired-tumor-normal-data">
<h3>Paired tumor-normal data<a class="headerlink" href="#paired-tumor-normal-data" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="tumor-only-datasets">
<h3>Tumor-only datasets<a class="headerlink" href="#tumor-only-datasets" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="quantifying-allelic-imbalance">
<h3>Quantifying allelic imbalance<a class="headerlink" href="#quantifying-allelic-imbalance" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="variant-detection">
<h3>Variant detection<a class="headerlink" href="#variant-detection" title="Permalink to this headline">¶</a></h3>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="resources.html" class="btn btn-neutral float-right" title="Resources" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="usage.html" class="btn btn-neutral float-left" title="Usage" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Katie Campbell.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>