Search.setIndex({docnames:["analysis","index","intro","references","resources","usage"],envversion:{"sphinx.domains.c":1,"sphinx.domains.changeset":1,"sphinx.domains.cpp":1,"sphinx.domains.javascript":1,"sphinx.domains.math":2,"sphinx.domains.python":1,"sphinx.domains.rst":1,"sphinx.domains.std":1,sphinx:54},filenames:["analysis.rst","index.rst","intro.rst","references.rst","resources.rst","usage.rst"],objects:{},objnames:{},objtypes:{},terms:{"0_beta":0,"4th":0,"class":0,"default":0,"export":0,"final":0,"float":0,"true":0,"while":0,The:0,Then:0,WES:0,abund:0,across:0,addit:1,albatross:[],align:0,all:0,all_typ:0,allel:1,analyz:1,ani:0,antigen:1,appropri:0,assign:[0,1],associ:0,base:0,been:5,bin:0,bone:[],call:[0,1],can:0,center:0,chosen:0,code:0,command:0,comprehens:0,construct:[1,5],contain:0,context:1,correspond:0,coverag:0,crunchi:[],custom:1,data:[1,5],databas:0,dataset:1,delimit:0,deriv:0,describ:[0,1],descript:0,design:1,desir:5,detect:1,differ:0,difficult:1,dna:5,docker:[0,1],document:[0,1],downstream:1,each:0,either:5,encompass:5,especi:1,exampl:0,extract:1,fals:0,field:0,field_of_resolut:0,file:0,filter:0,first:0,flag:0,follow:0,format:0,fq1:0,fq2:0,frog:[],from:[0,1],full:0,fullest:0,gannet:[],gen_msf_list:0,gene:[0,1],generate_reference_fil:0,genom:[0,1],genotyp:1,get:1,grep:0,haploid:1,hat:[0,5],have:[0,5],header:0,hisat2:1,hisat2_v2:0,hisat:1,hisat_prefix:0,hisatgenotype_extract_reads_v_kc:0,hisatgenotype_locu:0,hisatgenotype_locus_v_kc:0,hisatgenotype_modul:0,hisatgenotype_script:0,hla:5,hla_report_fil:0,hla_typ:0,hlahat_r:0,hlatyp:0,how:1,howev:0,human:1,identifi:0,imbal:1,implement:1,includ:[0,1],index:[],indic:0,inform:[0,1],input:5,integ:0,introduct:1,is_rna:0,kcampbel:0,keep:0,less:0,leukocyt:1,level:0,linear:1,list:0,locu:0,low:0,mai:0,make:1,manual:5,match:5,maximum:0,modul:[],modularli:5,most:0,name:0,natur:1,normal:[1,5],now:[],nuc_msf_list:0,one:0,onli:1,opt:0,option:0,other:5,out:[],output:0,outsid:0,page:[],pair:1,parallel:5,paramet:0,path:0,patient:0,perc_abund:0,percent:0,polymorph:1,process:1,pythonpath:0,quantifi:1,quantit:1,quantiti:[],rank:0,read:1,reduc:0,refer:1,region:0,rel:0,remain:0,remov:0,report:0,resolut:0,resourc:1,rippl:[],rna:[0,5],rnaseq:0,rnaseq_method:0,row:0,rscript:0,rubric:0,samtool:0,search:[],second:0,sep:0,sequenc:[0,1,5],set:0,sever:5,share:0,specif:1,start:1,stick:[],string:0,suffici:0,suggest:0,summar:0,support:1,tab:0,tabl:0,terra:1,than:0,thi:[0,1],third:0,those:0,thu:1,took:[],tool:5,top:0,top_hlatyp:0,treat:[],tsv:0,tumor:[1,5],two:0,txt:0,type:1,uniqu:0,upon:0,usag:1,used:[0,5],using:1,variant:1,visual:1,wdl:1,which:0,width:0,without:5,would:0,wouldn:[]},titles:["Data processing using HISAT2 and HISAT-genotype","HLA-HAT: HLA Haplotype Analysis Toolkit","Introduction","References","Resources","Usage"],titleterms:{The:1,allel:0,analysi:1,construct:0,custom:0,data:0,dataset:0,delight:[],detect:0,docker:5,document:[],extract:0,frozen:[],genotyp:0,guid:1,haplotyp:1,hat:1,hisat2:0,hisat:0,hla:[0,1],hlahat:[],imbal:0,indic:[],introduct:2,normal:0,onli:0,pair:0,process:0,quantifi:0,read:0,refer:[0,3],resourc:4,tabl:[],terra:5,toolkit:1,tumor:0,type:0,usag:5,user:1,using:0,variant:0,wdl:5,welcom:[]}})