---
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r hlahat_setup, include=FALSE, warning=FALSE, message=FALSE}
library(tidyverse)
library(data.table)
library(plotly)
library(DT)

```

## Allelic imbalance (DNA/RNA)
The HLA Haplotype Analysis Toolkit uses the ratio of sequencing coverage between tumor and normal to assess allelic imbalance at the DNA and RNA level in Class I and II HLA loci.

-  For tumor DNA, an allelic imbalance indicates a putative copy number aberration or loss-of-heterozygosity event at the gene or chromosomal level (see [HLA Copy Number][HLA Copy Number] for copy number and LOH assessment). 
-  For tumor RNA, an allelic imbalance indicates one HLA allele is expressed __relatively__ higher than the other.
```{r hlahat_stageinput}
types <- fread(find_types_f)
dna_snpld <- fread(dna_snpld_f)
dna_alleleld <- fread(dna_alleleld_f)
dna_locusld <- fread(dna_locusld_f)

rna_snpld <- fread(rna_snpld_f)
rna_alleleld <- fread(rna_alleleld_f)

classical_types <- c('A','B','C','DRA','DRB1','DQA1','DQB1','DPA1','DPB1')
sorted_hla_genes <- c('HFE','F','V','G','H',
                    'K','A','L','E','C',
                    'B','MICA','MICB','DRA','DRB1',
                    'DQA1','DQB1','DOB','TAP2','TAP1',
                    'DMB','DMA','DOA','DPA1','DPB1',
                    'DPB2')

purity <- allSols[sequenza_solution, ]$cellularity
```

### Summary

Allelic imbalance is measured using the using the 2^(log-fold difference (LD)) in sequencing coverage between alleles. 

- For DNA, allelic balance represents copy number differences at the DNA level. 
- For RNA, alleic imbalance represents differential expression of alleles. 

```{r hlahat_summary}
#Symbol | Category | Imbalance 
#-------|----------|----------------
#__~__  | Equal | 0-1.5
#__<__  | Low imbalance | 1.5-2.5
#__<<__ | Medium imbalance | 2.5-3.5
#__<<<__ | High imbalance | 3.5-Inf

alleleld_tab <- dna_alleleld %>% filter(Gene %in% LETTERS[1:3] & median_ld>0) %>% 
  mutate(`Allele 1` = CHROM, `Allele 2` = alt_gen) %>% 
  left_join(rna_alleleld, by = c('Gene','CHROM','alt_gen'), suffix = c('.dna','.rna')) %>% 
  mutate(degree_dna = case_when(2^abs(median_ld.dna) < 1.5 ~ "~", 
                                2^abs(median_ld.dna) < 2.5 ~ "<", 
                                2^abs(median_ld.dna) < 3.5 ~ "<<", 
                                TRUE ~ "<<<"),
         degree_rna = case_when(2^abs(median_ld.rna) < 1.5 ~ "~", 
                                2^abs(median_ld.rna) < 2.5 ~ "<", 
                                2^abs(median_ld.rna) < 3.5 ~ "<<", 
                                TRUE ~ "<<<"),
         degree_rna = ifelse(median_ld.rna > 0, degree_rna, gsub("<", ">", degree_rna))) %>%
  group_by(Gene) %>% 
  summarise(
    `DNA` = paste0(gsub("(^\\w\\*\\d+:\\d+):*.+", "\\1", `Allele 2`), " " , degree_dna, "  ", gsub("(^\\w\\*\\d+:\\d+):*.+", "\\1", `Allele 1`), " (", signif(2^median_ld.dna, 2), ")"),
    `RNA` = case_when(median_ld.rna > 0 ~ paste0(gsub("(^\\w\\*\\d+:\\d+):*.+", "\\1", `Allele 2`), " " , degree_rna, "  ", gsub("(^\\w\\*\\d+:\\d+):*.+", "\\1", `Allele 1`), " (", signif(2^median_ld.rna, 2), ")"),
                    TRUE ~ paste0(gsub("(^\\w\\*\\d+:\\d+):*.+", "\\1", `Allele 2`), " " , degree_rna, "  ", gsub("(^\\w\\*\\d+:\\d+):*.+", "\\1", `Allele 1`), " (", signif(2^median_ld.rna, 2), ")"))
            ) 

title = dt_caption("Allelic imbalance for DNA and RNA")
alleleld_tab %>%
  datatable(
    caption=title,
    extensions = dt_ext,
    options = dt_opts,
    rownames=FALSE,
    escape = FALSE
)
```
<font size="2"> * ~ equal, < low (1.5-25), << medium (2.5-3.5), <<< high (3.5-Inf)</font>

### Plots {.tabset}

#### DNA: AI vs Purity
The median allelic imbalance (AI) was calculated across all 26 HLA genes on Chromosome 6p and is summarized as a single value for the patient (shown by the red dot). The lines represent expected values of CNR based upon tumor purity.

Intersection between the point and the *Inf (LOH)* line indicates an extreme degree of imbalance and thus putative LOH.
```{r hlahat_dna_fulllocus}
genotypes2 <- c("1","12","112","1112","11112",
                "111112", "1111112", "11111112","11122")
purities2 <- seq(0, 1, 0.01)
exp_ald <- list(gen = genotypes2,
                 pur = purities2) %>% enframe %>% 
  spread(name, value) %>%
  unnest(gen) %>% unnest(pur) %>%
  group_by(gen) %>% mutate(n_1 = str_count(gen, "1"), n_2 = str_count(gen, "2")) %>% ungroup %>%
  mutate(ald = abs(log2(n_1) - log2(n_2)), cnr = 2^ald,
         pur_ald = abs(log2(pur*n_1/(n_1+n_2) + (1-pur)*0.5) - log2(pur*n_2/(n_1+n_2) + (1-pur)*0.5)),
         pur_cnr = 2^pur_ald) %>%
  arrange(-pur, cnr) %>% 
  mutate(gen = factor(gen, levels = unique(gen)),
         cnr = ifelse(cnr == Inf, "Inf (LOH)", cnr),
         cnr = factor(cnr, levels = unique(cnr)))

comp_topur <- ggplot(exp_ald, aes(y = pur_cnr, x = pur, group = gen, colour = factor(cnr))) +
  geom_line() +
  geom_point(data = dna_locusld, aes(x = purity, y = 2^median_ld, group = NA, fill = sid), colour = 'black', shape = 21, size = 3) +
  scale_colour_viridis_d(option = 'plasma') +
  scale_fill_manual(values = "Red") +
  coord_cartesian(ylim = c(1,10)) +
  scale_x_reverse() +
  scale_y_continuous(trans = 'log10') +
  labs(x = "Purity", y = "Allelic Imbalance", colour = "Expected AI", fill = "Sample", title = "Observed Alleic Imbalance vs. Tumor Purity")  +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5))

ggplotly(comp_topur)

```

#### DNA LD
The Log-fold Difference (LD) in depth of coverage at heterozygous SNPs are used to summarize the degree of allelic imbalance. 

- If there are equal copies of allele 1 and 2 (e.g. normal diploid), the LD value is 0. 
- The higher the LD, the larger the difference in coverage between the two alleles, indicating increased imbalance or LOH. 
- If there is LOH in the gene, the expected LD is infinity (unadjusted for tumor purity). 

Positive values for an allele indicate that the coverage was higher than its corresponding allele for that gene. The locus-level LD (median across all genes) is indicated by the grey dashed line. Each point represents a heterozygous SNP between allele 1 and allele 2, and the red lines indicate the median LD across all SNPs for the corresponding gene/allele. Genes are sorted by their position within the genome. 

Note that homozygous genes will not have any SNPs.
```{r hlahat_dna_allgenes_strip}
sort.dna_alleleld <- dna_alleleld %>% arrange(-median_ld) %>%
        mutate(CHROM = factor(CHROM, levels = CHROM),
               Gene = factor(Gene, levels = sorted_hla_genes)) %>% 
  group_by(Gene) %>% mutate(alleleN = 1:n())
sort.dna_snpld <- dna_snpld %>% 
        mutate(CHROM = factor(CHROM, levels = levels(sort.dna_alleleld$CHROM)),
               Gene = factor(Gene, levels = sorted_hla_genes)) %>%
  left_join(sort.dna_alleleld)
gene_colors <- data.frame(Gene = sorted_hla_genes) %>% 
  mutate(color = case_when(Gene %in% classical_types ~ "red", TRUE ~ "grey50"))

dna_imb <- ggplot(data = NULL, aes(x = Gene, colour = factor(alleleN), label = CHROM)) +
  geom_hline(yintercept = 0) +
        geom_hline(yintercept = c(-dna_locusld$median_ld, dna_locusld$median_ld), linetype = 'dashed', colour = 'black', alpha = 0.5) +
        geom_jitter(data = filter(sort.dna_snpld, !is.na(CHROM)), aes(y = ld), width = 0.2, alpha = 0.5) +
        geom_point(data = filter(sort.dna_alleleld, !is.na(CHROM)), aes(y = median_ld), shape = 95, colour = 'red', size = 4) +
        scale_y_continuous(sec.axis = sec_axis(~ 2 ^ ., name = "Copy Number Ratio", breaks = c(1,2,4,8)), breaks = seq(-20, 20, 1)) +
        scale_x_discrete(drop = FALSE) +
        scale_colour_viridis_d() +
        labs(x = "Gene", y = "Log-fold Difference in Coverage", title = "Allelic Imbalance of HLA Alleles",
        colour = "Allele N") +
        theme_bw() +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, color = gene_colors$color),
              plot.title = element_text(hjust = 0.5))
ggplotly(dna_imb)
```

#### DNA LD Class I
The LD value per SNP (each point) is shown for each allele for the 3 classical Class I alleles, according to their genomic position.
```{r hlahat_dnaclassI_strip}
genes <- c('A','B','C')
dna_imbC <- ggplot(data = NULL) +
  geom_hline(data = filter(sort.dna_alleleld, Gene %in% genes), aes(yintercept = median_ld, colour = factor(CHROM), label = CHROM), alpha = 0.5) +
#  geom_hline(data = filter(sort.dna_alleleld, Gene %in% genes), aes(yintercept = ld_25, colour = factor(CHROM), label = CHROM), linetype = 'dashed', alpha = 0.5) +
#  geom_hline(data = filter(sort.dna_alleleld, Gene %in% genes), aes(yintercept = ld_75, colour = factor(CHROM), label = CHROM), linetype = 'dashed', alpha = 0.5) +
  geom_point(data = filter(sort.dna_snpld, Gene %in% genes & alleleN == 1), aes(x = END, y = ld, colour = factor(CHROM), label = CHROM)) +
  geom_point(data = filter(sort.dna_snpld, Gene %in% genes & alleleN == 2), aes(x = alt_pos, y = ld, colour = factor(CHROM), label = CHROM)) +
  scale_colour_viridis_d() +
  facet_grid(. ~ Gene, scales = 'free_x', space = 'free_x') +
  labs(x = "Position within Gene", colour = "Alelle", title = "Allelic Imbalance of Class I Alleles",
       y = "Log-fold Difference in Coverage") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5))

ggplotly(dna_imbC)
```

#### RNA LD
The Log-fold Difference (LD) in depth of coverage at heterozygous SNPs are used to summarize the degree of allelic imbalance. 

- If there are equal copies of allele 1 and 2 (e.g. normal diploid), the LD value is 0. 
- The higher the LD, the larger the difference in coverage between the two alleles, indicating increased imbalance or LOH. 
- If there is LOH in the gene, the expected LD is infinity (unadjusted for tumor purity). 

Positive values for an allele indicate that the coverage was higher than its corresponding allele for that gene. The locus-level LD (median across all genes) is indicated by the dashed line. Each point represents a heterozygous SNP between allele 1 and allele 2, and the red lines indicate the median LD across all SNPs for the corresponding gene/allele. Genes are sorted by their position within the genome. 

Note that homozygous genes will not have any SNPs.
```{r hlahat_rna_allgenes_scatter}
sort.rna_alleleld <- rna_alleleld %>% arrange(median_ld) %>%
         mutate(CHROM = factor(CHROM, levels = CHROM),
               Gene = factor(Gene, levels = sorted_hla_genes)) %>%
  left_join(sort.dna_alleleld, by = c("Gene", "CHROM", "alt_gen"), suffix = c(".rna", ".dna"))
sort.rna_snpld <- rna_snpld %>%
         mutate(CHROM = factor(CHROM, levels = levels(sort.rna_alleleld$CHROM)),
               Gene = factor(Gene, levels = sorted_hla_genes)) %>%
  left_join(sort.dna_alleleld)
gene_colors <- data.frame(Gene = sort.rna_alleleld$Gene) %>% mutate(color = case_when(Gene %in% classical_types ~ "red", TRUE ~ "grey50"))

rna_imb <- ggplot(data = NULL, aes(x = Gene, colour = factor(alleleN), label = CHROM)) +
        geom_hline(yintercept = c(-dna_locusld$median_ld, dna_locusld$median_ld), linetype = 'dashed', colour = 'black', alpha = 0.5) +
        geom_jitter(data = filter(sort.rna_snpld, !is.na(alleleN)), aes(y = ld), width = 0.2, alpha = 0.5) +
        geom_point(data = filter(sort.rna_alleleld, !is.na(alleleN)), aes(y = median_ld.rna), shape = 95, colour = 'red', size = 4) +
        scale_y_continuous(#sec.axis = sec_axis(~ 2 ^ ., name = "Copy Number Ratio", breaks = c(1,2,4,8)), 
                           breaks = seq(-20, 20, 1)) +
        scale_x_discrete(drop = FALSE) +
        scale_colour_viridis_d(drop = TRUE) +
        labs(x = "Gene", y = "Log-fold Difference in Coverage", title = "Expression Imbalance of HLA Alleles",
             colour = "Allele N") +
        theme_bw() +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, color = gene_colors$color),
              plot.title = element_text(hjust = 0.5))

ggplotly(rna_imb)
```

#### RNA LD Class I
The LD value per SNP (each point) is shown for each allele for the 3 classical Class I alleles, according to their genomic position.
```{r hlahat_rna_imbC}
rna_imbC <- ggplot(data = NULL) +
  geom_hline(data = filter(sort.rna_alleleld, Gene %in% genes), aes(yintercept = median_ld.rna, colour = factor(CHROM), label = CHROM), alpha = 0.5) +
#  geom_hline(data = filter(sort.rna_alleleld, Gene %in% genes), aes(yintercept = ld_25.rna, colour = factor(CHROM), label = CHROM), linetype = 'dashed', alpha = 0.5) +
#  geom_hline(data = filter(sort.rna_alleleld, Gene %in% genes), aes(yintercept = ld_75.rna, colour = factor(CHROM), label = CHROM), linetype = 'dashed', alpha = 0.5) +
  geom_point(data = filter(sort.rna_snpld, Gene %in% genes & alleleN == 1), aes(x = END, y = ld, colour = factor(CHROM), label = CHROM)) +
  geom_point(data = filter(sort.rna_snpld, Gene %in% genes & alleleN == 2), aes(x = alt_pos, y = ld, colour = factor(CHROM), label = CHROM)) +
  scale_colour_viridis_d() +
  facet_grid(. ~ Gene, scales = 'free_x', space = 'free_x') +
  labs(x = "Position within Gene", colour = "Allele", y = "Log-fold Difference in Coverage",
       title = "Expression Imbalance of Class I Alleles") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5))

ggplotly(rna_imbC)
```

#### DNA v. RNA
Alleic imbalance in RNA expression may be associated with allelic imbalance at the DNA level (i.e. copy number variation). 

Genes that fall along the positive diagonal line indicate differential allele expression is strongly associated with the number of copies of the allele at the DNA level.
```{r hlahat_dnavrna_scatter}
comp_dnaRna <- ggplot(sort.rna_alleleld, 
       aes(x = median_ld.dna, y = median_ld.rna, colour = Gene, label = CHROM)) +
  geom_abline(linetype = 'dotted', colour = 'grey80') +
  geom_abline(slope = -1, linetype = 'dotted', alpha = 0.5) +
  geom_point() +
  scale_colour_viridis_d() +
  coord_cartesian(xlim = c(min(sort.dna_alleleld$median_ld), max(sort.dna_alleleld$median_ld)),
                  ylim = c(min(filter(sort.rna_alleleld, median_ld.rna>-Inf)$median_ld.rna), max(filter(sort.rna_alleleld, median_ld.rna<Inf)$median_ld.rna))) +
  labs(x = "Log-fold Difference in Coverage- DNA", y = "Log-fold Difference in Coverage- RNA", colour = "Gene",
       title = "Allele Coverage in RNA v. DNA") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5))

ggplotly(comp_dnaRna)
```

### HLA types
```{r hlahat_haplotypes_tab}
hlahat_types_tab <- types %>% 
  mutate(Classical = case_when(gene %in% classical_types ~ "Classical", TRUE ~ "Other"),
                 Classical = factor(Classical, levels = c('Classical','Other')),
                 gene = factor(gene, levels = sorted_hla_genes)) %>% 
        arrange(Classical, gene) %>%
        dplyr::select(Gene=gene, hla_call, alleleN, Classical) %>% 
        spread(alleleN, hla_call) 

title = dt_caption("HLA types for HLA Class I and II loci")
hlahat_types_tab %>%
  datatable(
    caption=title,
    extensions = dt_ext,
    options = dt_opts,
    rownames=FALSE,
    escape = FALSE
)
```

### Descriptive statistics {.tabset}

#### DNA LD
```{r hlahat_dnald_tab}
dna_alleleld_tab <- dna_alleleld
colnames(dna_alleleld_tab) <- c('Gene', 'Allele 1', 'Allele 2', 'Min LD', 'Median LD', 'Max LD', 'SNPs')
dna_alleleld_tab <- dna_alleleld_tab %>%
  relocate(SNPs, .after = `Allele 2`) %>%
  dplyr::filter(`Median LD` > 0) %>%
  mutate_if(is.numeric, round, 2) %>%
  mutate(SNPs = as.integer((SNPs)))
  
title = dt_caption("DNA log fold difference (LD) descriptive statistics")
dna_alleleld_tab %>%
  datatable(
    caption=title,
    extensions = dt_ext,
    options = dt_opts,
    rownames=FALSE,
    escape = FALSE
) 
```

#### RNA LD
```{r hlahat_rnald_tab}
rna_alleleld_tab <- rna_alleleld
colnames(rna_alleleld_tab) <- c('Gene', 'Allele 1', 'Allele 2', 'Min LD', 'Median LD', 'Max LD', 'SNPs')
rna_alleleld_tab <- rna_alleleld_tab %>%
  relocate(SNPs, .after = `Allele 2`) %>%
  dplyr::filter(`Median LD` > 0) %>%
  mutate_if(is.numeric, round, 2) %>%
  mutate(SNPs = as.integer(SNPs))

title = dt_caption("RNA log fold difference (LD) descriptive statistics")
rna_alleleld_tab %>%
  datatable(
    caption=title,
    extensions = dt_ext,
    options = dt_opts,
    rownames=FALSE,
    escape = FALSE
) 
```
